### PR TITLE
[aot] FindTaichi CMake module to help outside project integration

### PR DESCRIPTION
**Pull request by PENGUINLIONG**
Issue: #

### Brief Summary

This PR introduces a CMake module `FindTaichi.cmake` to help external project to integrate Taichi Runtime into their applications following the CMake convention to minimize the lines of codes they need in their `CMakeLists.txt`.

The following is the integration CMake snippet:

```cmake
# We need to set `FindTaichi.cmake` in module path to `find_package(taichi)`.
set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../cmake)
find_package(taichi REQUIRED)

add_executable(${EXAMPLE_NAME} main.cpp)
target_link_libraries(${EXAMPLE_NAME} PRIVATE taichi::runtime)

# Copy redistributed libraries (`.so` and `.dll`) next to the executable.
add_custom_command(
    TARGET ${EXAMPLE_NAME}
    PRE_BUILD
    COMMENT "-- Copy redistributed libraries to output directory (if different): ${taichi_REDIST_LIBRARIES}"
    COMMAND ${CMAKE_COMMAND} -E copy_if_different ${taichi_REDIST_LIBRARIES} "$<TARGET_FILE_DIR:${EXAMPLE_NAME}>")
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,11 +8,13 @@ project(taichi)
 
 include("cmake/utils.cmake")
 
+message(WARNING "It seems that you are running cmake manually, which may cause issues. Please use setup.py to build taichi from source, see https://docs.taichi-lang.org/docs/dev_install for more details.")
```
Comment (ailzhang) [inline CMakeLists.txt:11]: nit: why do we take this out of the if branch? ;) The old logic was: if we run cmake from setup.py it'll have `TI_VERSION_MAJOR` set so it won't trigger this warning, not sure if it's still applicable.

**Review by ailzhang — COMMENTED** [review]
lgtm! leaving this to @jim19930609 for a final check ;)

File: CMakeLists.txt (hunk 1 above)
Comment (PENGUINLIONG) [reply CMakeLists.txt]: oof it's a miss

File: cmake/TaichiCAPI.cmake [hunk 2]
```diff
@@ -32,14 +45,17 @@ endif()
 
 if (TI_WITH_OPENGL)
   list(APPEND C_API_SOURCE "c_api/src/taichi_opengl_impl.cpp")
+  list(APPEND C_API_PUBLIC_HEADERS "c_api/include/taichi/taichi_opengl.h")
 endif()
 
 if (TI_WITH_METAL)
   list(APPEND C_API_SOURCE "c_api/src/taichi_metal_impl.mm")
+  #list(APPEND C_API_PUBLIC_HEADERS "c_api/include/taichi/taichi_metal.h")
```
Comment (jim19930609) [inline cmake/TaichiCAPI.cmake:53]: uncomment this?

File: c_api/cmake/FindTaichi.cmake [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+# Support `IN_LIST` in CMake `if` command.
+if(POLICY CMP0057)
+    cmake_policy(SET CMP0057 NEW)
+endif()
+
+# Find TiRT in the installation directory. The installation directory is
+# traditionally specified by an environment variable
+# `TAICHI_C_API_INSTALL_DIR`.
+if(NOT EXISTS ${TAICHI_C_API_INSTALL_DIR})
```
Comment (jim19930609) [inline c_api/cmake/FindTaichi.cmake:63]: Will `FindTaichi.cmake` get shipped along with C-API? If so,  ${TAICHI_C_API_INSTALL_DIR} should be a relative path to `FindTaichi.cmake` (like `./..`?)

**Review by jim19930609 — APPROVED** [review]

File: cmake/TaichiCAPI.cmake (hunk 2 above)
Comment (PENGUINLIONG) [reply cmake/TaichiCAPI.cmake:53]: We don't have this file atm, lol.

File: c_api/cmake/FindTaichi.cmake (hunk 3 above)
Comment (PENGUINLIONG) [reply c_api/cmake/FindTaichi.cmake:63]: This one should be set up by the user. This FindTaichi.cmake is expected to be copied by the user to their repositories, in a folder called `cmake`. Then, when the user builds their project, CMake can find that package using this script following the rules in `https://cmake.org/cmake/help/v3.0/command/find_package.html#command:find_package`.

**Pull request merged**